### PR TITLE
Feature/cgroup resource management

### PR DESCRIPTION
**Pull request by colindrewes**
This PR adds support for joining cgroups and applying resource quota with cgroups. Resource quotas can be applied to the cpu, memory, and io cgroups. 

To use the cgroup feature, an existing cgroup must be accessible on the fs. We will use the default mount point /sys/fs/cgroup in the following examples. If the cgroup hierarchy is not mounted at the default location /sys/fs/cgroup, it can be mounted with `sudo mount -t cgroup2 cgroup /sys/fs/cgroup`. To create a cgroup with the appropriate permissions for bst to join or apply resource quotas from, first ensure that `/sys/fs/cgroup/cgroup.subtree_control` contains whatever controller that the bst user intends to use (i,.e. if you wish for bst to control cpu then ensure that `cat /sys/fs/cgroup/cgroup.subtree_control` contains "cpu". If not, it can be added with `echo +cpu | sudo tee /sys/fs/cgroup/cgroup.subtree_control`). The `/sys/fs/cgroup/bst` hierarchy must also at the very least have the pids controller so that bst can add and remove itself from the cgroup. This can be accomplished with `echo +pids | sudo tee /sys/fs/cgroup/cgroup.subtree_control`. Now one can create a cgroup that bst will operate within using `sudo mkdir /sys/fs/cgroup/bst`. The sub-hierarchy can be deletehgated to the bst user with `sudo chown -R $USER:$USER /sys/fs/cgroup/bst`. The bst user will also need to be able to add itself to the `/sys/fs/cgroup/bst/cgroup.procs` which requires that the user have write permissions on `/sys/fs/cgroup/cgroup.procs` which can be accomplished with `sudo chown $USER:$USER /sys/fs/cgroup/cgroup.procs`. 

If the user then manually applies cgroup quotas `/sys/fs/cgroup/bst`, it can be joined with `bst --cgroup=/sys/fs/cgroup/bst`. The bst instance will be automatically added to the cgroup and whatever quotas have been specified by the user will be applied. 

If the user wishes for bst to configure quotas within the new bst cgroup, that can be accomplished with the `--climit <resource>=<limit>`. This still depends on the user specifying some root cgroup with the --cgroup=<path> that bst will be able to access (i.e. simply running `bst --climit cpu.max=5000` will not apply any limits, while running `bst --cgroup=/sys/fs/cgroup/ --climit cpu.max=5000` will limit cpu usage). This PR supports the following limits:
- `bst --cgroup=/sys/fs/cgroup/bst --climit cpu.max=$MAX` -- This will control the $MAX bandwidth limit for each $PERIOD duration. The $PERIOD duration is 100000 by default. 
- `bst --cgroup=/sys/fs/cgroup/bst --climit cpu.max="$MAX $PERIOD"` -- This will control the $MAX bandwidth limit for each $PERIOD duration. 
- `bst --cgroup=/sys/fs/cgroup/bst --climit cpu.weight=$WEIGHT` -- This will control the cpu proportional weight of the bst process within the parent cgroup. A parent's resource (/sys/fs/cgroup/bst in this case), is distributed by "adding up the weights of all active children and giving each the fraction matching the ratio of its weight against the sum" [[ref](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html)]. This value is in the range [1,1000].
- `bst --cgroup=/sys/fs/cgroup/bst --climit memory.min=$MIN` -- when bst memory usage is within min boundary the memory won't be reclaimed under any conditions. This can use postfixes like G, M, etc.
- `bst --cgroup=/sys/fs/cgroup/bst --climit memory.low=$LOW` -- similar to memory.min but soft limit.
- `bst --cgroup=/sys/fs/cgroup/bst --climit memory.high=$HIGH` -- memory usage throttle limit. As will memory bounds, this can use postfixes.
- `bst --cgroup=/sys/fs/cgroup/bst --climit memory.max=$MAX` -- memory hard limit.
- `bst --cgroup=/sys/fs/cgroup/bst --climit memory.swap.high=$HIGH` -- swap usage throttle limit.
- `bst --cgroup=/sys/fs/cgroup/bst --climit memory.swap.max=$MAX` -- swap usage hard limit.
- `bst --cgroup=/sys/fs/cgroup/bst --climit io.weight=$WEIGHT` -- updates default weight which is a number [1,1000] that specifies the relative amount of IO time the bst instance will use in relation to its siblings. 
-`bst --cgroup=/sys/fs/cgroup/bst --climit io.max="$MAJ:$MIN rbps=$RIOPS wbps=$WBPS riops=$RIOPS wiops=$WIOPS"` -- Sets the max IO usage for the device specified by the major minor code. 
-`bst --cgroup=/sys/fs/cgroup/bst --climit io.latency="$MAJ:$MIN target=$MS"` -- Set the number of ms a process can wait before IO from other procs must be given to it. 
-`bst --cgroup=/sys/fs/cgroup/bst --climit pids.max=$MAX` -- Max number of processes that can be controlled by the bst cgroup.

One could use multiple in conjunction: 

`bst --cgroup=/sys/fs/cgroup/bst --climit cpu.max="5000 7000" --climit cpu.weight=100 --climit memory.max=1G --climit memory.low=1M --climit memory.max=1G --climit memory.swap.max=1G --climit io.max="8:0 rbps=100 wbps=100 riops=100 wiops=100" --climit io.latency="8:0 target=20" --climit pids.max=max`
 
for fine granularity resource control on bst. 

To provide a consistent view of the cgroup hierarchy from within bst, if /sys/fs/cgroup is a cgroups mount point it will be mounted over (tmpfs mount + cgroup mount) which will render the bst cgroup (the root cgroup within bst) as the cgroup of the system. If /sys/fs/cgroup is unmounted or bst is operating from a different rootfs nothing will be done. As well as this, bst will not mount over /sys/fs/cgroup if one of the following flags is in place:
- --no-cgroup-remount
- --share cgroup

If this all looks appropriate I can write the corresponding entries in the manual.

### CODE REVIEW

File: mount.c [hunk 1]
```diff
@@ -160,12 +162,16 @@ void mount_entries(const char *root, const struct mount_entry *mounts, size_t nm
...
 		   mounting something with the same source and target as one of our
 		   default mounts. Right now, the only case where this applies is the
-		   automatic /proc remount. */
+		   automatic /proc remount. TODO UPDATE THIS???*/
 		size_t targetlen = strlen(target);
 		if (strcmp(mnt->source, "proc") == 0 && strcmp(target + targetlen - 5, "/proc") == 0) {
 			umount2(target, MNT_DETACH);
 		}
-
+    // TODO This is maybe needed here?
```
Comment (Snaipe) [inline mount.c]: Yes, I'd assume so. Otherwise, an explicit `bst --mount cgroup,/sys/fs/cgroup,cgroup2` would fail for no reason.

File: enter.c [hunk 2]
```diff
@@ -225,13 +226,65 @@ int enter(struct entry_settings *opts)
 		}
 	}
 
+	int rootpid = getpid();
+	bool req_cgroup_clean = false;
+	int cleanfd;
+
+	// Do not join the root cgroup if limits are to be applied, otherwise bst needs to create sub-hierarchies
+	if (!opts->nactiveclimits && opts->cgroup.present) {
```
Comment (Snaipe) [inline enter.c:234]: It's better to be explicit in intent for quantities -- I'd use `if (opts->nactiveclimits > 0 && opts->cgroup.present)`. This ensures that this stays correct even if the type of the quantity changes to some signed value for whatever reason. Explicit is better than clever, and there has been many situations where bugs have been introduced because of the implicit meaning of nonzero to mean true.

File: enter.c [hunk 3]
```diff
@@ -225,13 +226,65 @@ int enter(struct entry_settings *opts)
...
+
+		int rootfd = open(root, O_PATH, 0);
+		if (rootfd == -1) {
+			 err(1, "open %s", root);
+		}
+
+		const char * rootcgroup = opts->cgroup.cgroup_path;
+		const char * subcgroup = makepath("%s/bst.%d", rootcgroup,rootpid);
+
+		if (mkdir(subcgroup, 0755) != -1) {
```
Comment (Snaipe) [inline enter.c:254]: This should be 0777 -- don't chose the mode for the user, let their umask choose.

File: enter.c [hunk 4]
```diff
@@ -225,13 +226,65 @@ int enter(struct entry_settings *opts)
...
+
+	} else if (opts->nactiveclimits && opts->cgroup.present) {
+
+		int rootfd = open(root, O_PATH, 0);
+		if (rootfd == -1) {
+			 err(1, "open %s", root);
+		}
+
+		const char * rootcgroup = opts->cgroup.cgroup_path;
+		const char * subcgroup = makepath("%s/bst.%d", rootcgroup,rootpid);
```
Comment (Snaipe) [inline enter.c:252]: micro-nit: missing space after `rootcgroup,`.

I should probably look into putting some uncrustify config for code formatting.

File: enter.c [hunk 5]
```diff
@@ -225,13 +226,65 @@ int enter(struct entry_settings *opts)
...
+			int cgroupfd = open(subcgroup, O_PATH | O_DIRECTORY);
+			if (cgroupfd == -1) {
+				err(1, "unable open cgroup %s", subcgroup);
+			}
+
+			join_cgroup(subcgroup, cgroupfd);
+			close(cgroupfd);
+
+			// The creation of this sub-hierarchy will require later cleanup
+			req_cgroup_clean = true;
```
Comment (Snaipe) [inline enter.c:273]: How about initializing cleanfd to -1, and use this as a condition for determining whether it needs cleanup or not?

File: cgroups.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		err(1, "joining cgroup %s", cgroup_path);
+	}
+	close(fd);
+}
+
+/* Called when bst terminates. This removes the created bst directory at the given
+ * cleanfd file descriptor.
+ */
+void cgroup_clean(int cleanfd, const char * cgroup_clean) {
+	join_cgroup(cgroup_clean, cleanfd);
```
Comment (Snaipe) [inline cgroups.c:76]: why is this needed?

File: enter.c [hunk 7]
```diff
@@ -468,6 +527,45 @@ int enter(struct entry_settings *opts)
...
+			err(1, "fstat %s", root);
+		}
+
+		close(rootfd);
+
+		// Is /sys/fs/cgroup already a mountpoint?
+		if (cgroupst.st_dev != 0 && cgroupst.st_dev != rootst.st_dev) {
+			const char *target = makepath("%s/sys/fs/cgroup", root);
+
+			// Mount tmpfs over /sys/fs/cgroup
```
Comment (Snaipe) [inline enter.c:557]: Comments should explain the why, not the what, otherwise they end up redundant -- I can read the mount("tmpfs", target, "tmpfs", 0, NULL) and gather the exact same information as your comment.

What would be more interesting is explaining why mounting a tmpfs over the destination is necessary.

File: enter.c [hunk 8]
```diff
@@ -468,6 +527,45 @@ int enter(struct entry_settings *opts)
...
+			if (mount("tmpfs", target, "tmpfs", 0, NULL) == -1) {
```
Comment (Snaipe) [inline enter.c:558]: Speaking of, why not do the same kind of stuff as the /proc remount? Having an extra tmpfs mount is a bit more expensive than just going the extra mile and handling the umount case correctly.

File: enter.c (hunk 5 above)
Comment (colindrewes) [reply enter.c]: Good thought.

File: cgroups.c (hunk 6 above)
Comment (colindrewes) [reply cgroups.c]: I have added a comment explaining this. bst cannot remove a cgroup if the bst pid is a member of that cgroup. This will give some "device or resource busy" error if one tries otherwise. Open to other ideas here though.

File: enter.c (hunk 7 above)
Comment (colindrewes) [reply enter.c]: Added clarifying comments based on our discussion.

File: enter.c (hunk 8 above)
Comment (colindrewes) [reply enter.c:565]: Unfortunately the same flow did not work for sys/fs/cgroup as /proc. Just leaving this comment here for posterity.

File: man/bst.1.scd [hunk 9]
```diff
@@ -229,6 +229,42 @@ _VAR=value_ before the executable to run.
 
 	Supported options are described in more detail in the *NETWORKING* section.
 
+\--cgroup <path>
+	Specify the cgroup directory that bst will join or operate within. 
+
+	If no climits are provided, then bst will join the cgroup at the provided <path>.
+	If climits are provided then bst will create cgroup sub-hierarchies at the <path>.
```
Comment (Snaipe) [inline man/bst.1.scd]: I think it might be more accurate to say that bst will always join the specified cgroup, and that it will create subhierarchies as necessary.

File: test/bst.t [hunk 10]
```diff
@@ -245,3 +245,187 @@ Testing Environment
...
+	40
+
+	$ bst --cgroup /sys/fs/cgroup/bst --no-cgroup-remount --climit pids.max=19 cat /sys/fs/cgroup/bst/pids.max
+	19
+
+	$ ! bst --cgroup /sys/fs/cgroup/bst --no-cgroup-remount --climit pids.max=-1 cat /sys/fs/cgroup/bst/pids.max 2>/dev/null
+
+Test mutliple climits in conjunction
+
+	$ bst --cgroup=/sys/fs/cgroup/bst --no-cgroup-remount --climit cpu.max="5000 7000" --climit cpu.weight=100 --climit memory.max=1G --climit memory.low=1M --climit memory.max=1G --climit memory.swap.max=1G --climit io.max="8:0 rbps=100 wbps=100 riops=100 wiops=100" --climit io.latency="8:0 target=20" --climit pids.max=50 cat /sys/fs/cgroup/bst/cpu.max; cat /sys/fs/cgroup/bst/cpu.weight; cat /sys/fs/cgroup/bst/memory.min; cat /sys/fs/cgroup/bst/memory.low; cat /sys/fs/cgroup/bst/memory.high; cat /sys/fs/cgroup/bst/memory.max; cat /sys/fs/cgroup/bst/memory.swap.high; cat /sys/fs/cgroup/bst/memory.swap.max; cat /sys/fs/cgroup/bst/io.weight; cat /sys/fs/cgroup/bst/io.max; cat /sys/fs/cgroup/bst/io.latency; cat /sys/fs/cgroup/bst/pids.max
```
Comment (Snaipe) [inline test/bst.t]: how about breaking this into multiple lines:

```
$ bst --cgroup=/sys/fs/cgroup/bst \
>   --no-cgroup-remount \
>   --climit cpu.max="5000 7000" \
>   --climit cpu.weight=100 \
>   --climit memory.max=1G \
>   --climit memory.low=1M \
>   --climit memory.max=1G \
>   --climit memory.swap.max=1G \
>   --climit io.max="8:0 rbps=100 wbps=100 riops=100 wiops=100" \
>   --climit io.latency="8:0 target=20" \
>   --climit pids.max=50 \
>   cat /sys/fs/cgroup/bst/cpu.max
> cat /sys/fs/cgroup/bst/cpu.weight
> cat /sys/fs/cgroup/bst/memory.min
> cat /sys/fs/cgroup/bst/memory.low
> cat /sys/fs/cgroup/bst/memory.high
> cat /sys/fs/cgroup/bst/memory.max
> cat /sys/fs/cgroup/bst/memory.swap.high
> cat /sys/fs/cgroup/bst/memory.swap.max
> cat /sys/fs/cgroup/bst/io.weight
> cat /sys/fs/cgroup/bst/io.max
> cat /sys/fs/cgroup/bst/io.latency
> cat /sys/fs/cgroup/bst/pids.max
```

Also note that only the first cat is being executed in bst, in this current form. What you can do instead is simply pass --workdir /sys/fs/cgroup, then follow that with `cat cpu.max cpu.weight memory.min memory.low ... pids.max`.

File: test/bst.t [hunk 11]
```diff
@@ -245,3 +245,187 @@ Testing Environment
...
+
+	$ if touch /sys/fs/cgroup/cgroup.procs         ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst                  ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/cpu.max          ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/cpu.weight       ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/memory.min       ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/memory.low       ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/memory.high      ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/memory.max       ; then exit 0; else exit 80; fi
+	$ if touch /sys/fs/cgroup/bst/memory.swap.high ; then exit 0; else exit 80; fi
```
Comment (Snaipe) [inline test/bst.t:259]: This will likely make the tests never run, ever. A lot of systems don't configure swap, especially when they have ssds/nvmes, since swap tends to kill these disks faster.

It's important that the tests run at the very least in the CI process.

**Review by Snaipe — CHANGES_REQUESTED** [review]
This is better, a couple of nits. There's a more critical flaw however, in that there is no way to use this feature without making the system less secure, because it's necessary to get write access to the parent cgroup.procs.

File: test/bst.t (hunk 11 above)
Comment (colindrewes) [reply test/bst.t]: I lowered the scope of the tests to what should run on most platforms.

File: test/bst.t (hunk 10 above)
Comment (colindrewes) [reply test/bst.t]: I had some issues using multi-line inputs in the bst test suite. Is there some special syntax needed here?

File: test/bst.t [hunk 12]
```diff
@@ -245,3 +245,187 @@ Testing Environment
...
+	$ ! bst --cgroup /sys/fs/cgroup/bst test -f /sys/fs/cgroup/bst/cgroup.procs
+
+Test trying to apply climits without specifying a cgroup
+
+	$ ! bst --no-cgroup-remount --climit cpu.max=3333 cat /sys/fs/cgroup/bst/cpu.max
+	bst: unable to apply limits without --cgroup specified: Success
+
+Test cpu.max cgroup usage
+
+	$ bst --cgroup /sys/fs/cgroup/bst --no-cgroup-remount --climit cpu.max="max 100000" cat /sys/fs/cgroup/bst/cpu.max
```
[the remark this answers is not in the document]
Comment (colindrewes) [reply test/bst.t]: The issue I encountered here was that the --limits are applied in the /sys/fs/cgroup/bst not to the /sys/fs/cgroup/bst.<pid> which is remounted as /sys/fs/cgroup. Any ideas here?

File: enter.c [hunk 13]
```diff
@@ -153,6 +153,35 @@ int enter(struct entry_settings *opts)
...
+
+	// If this set (!= -1), then it is the open fd of bst directory needed to be cleaned
+	int cleanfd = -1;
+
+	/*
+	 * If there is a cgroup specified we need to create bst.<pid> subcgroup to join
+	 * Otherwise if no cgroup is specified but limits are applied bst will error out
+	 */
+	if (opts->cgroup_path != NULL) {
+		cleanfd = open(opts->cgroup_path, O_PATH | O_DIRECTORY | O_CLOEXEC);
```
Comment (Snaipe) [inline enter.c:168]: This is a bit delicate. All of the code prior to `setuid(getuid())` on L207 is privileged, and may be running as setuid root. This means that the open will always succeed.

The only way this can be secure is if you either open the file after dropping privileges and then send the file descriptor to the helper, or leave the current code as-is and perform an explicit permission check. The easier (and more correct) way to do this is the former, since the latter requires you to emulate the kernel permission model, and we're going to get it wrong (checking user & mode is insufficient, and ignores acls, selinux and apparmor, seccomp, ...).

I have some pending changes to allow this as part of my seccomp supervisor. I'll cherry-pick some of the commits that does this, and you'll just be able to call send_fd.

File: outer.c [hunk 14]
```diff
@@ -207,6 +263,21 @@ static void persist_ns_files(pid_t pid, const char **persist)
...
+	if (helper->cleanfd != -1) {
+		pid_t pid = fork();
+		if (pid == -1) {
+			err(1, "outer_helper: cgroup cleanup fork");
+		}
+
+		if (pid == 0) {
+			cgroup_helper(helper->cleanfd, rootpid);
+			_exit(0);
+		}
```
Comment (Snaipe) [inline outer.c:278]: I'd maybe leave a comment saying that we explicitly let the process leak in order for the cleanup to happen after bst exits.

File: outer.c [hunk 15]
```diff
@@ -183,6 +185,60 @@ static void persist_ns_files(pid_t pid, const char **persist)
...
+
+	char populated[BUFSIZ];
+	int ready = 0;
+
+	do {
+		// Single event (update cgroup.procs), block indefinitely
+		ready = epoll_wait(epollfd, &event, 1, -1);
+
+		// We need a new fd to read from the cgroup.events
+		int readfd = openat(cgroupfd, "cgroup.events", O_RDONLY);
```
Comment (Snaipe) [inline outer.c:229]: It's likely easier to instead seek to the beginning of the file. Also, this needs error handling.

File: outer.c [hunk 16]
```diff
@@ -183,6 +185,60 @@ static void persist_ns_files(pid_t pid, const char **persist)
...
 
+/*
+ * If bst has entered a cgroup this function will epoll the cgroup.events file
+ * to detect when all pids have exited the cgroup ("populated 0"). The cgroup is 
+ * destroyed when this condition is met.
+ */
+static void cgroup_helper(int cleanfd, pid_t rootpid)
+{
+	char *subcgroup = makepath("bst.%d", rootpid);
+	int cgroupfd = openat(cleanfd, subcgroup, O_DIRECTORY);
```
Comment (Snaipe) [inline outer.c:196]: error handling is missing

File: outer.c [hunk 17]
```diff
@@ -183,6 +185,60 @@ static void persist_ns_files(pid_t pid, const char **persist)
...
+	if (epoll_ctl(epollfd, EPOLL_CTL_ADD, cevent, &event) == -1) {
+		err(1, "epoll_ctl_add cgroupfd");
+	}
+
+	char populated[BUFSIZ];
+	int ready = 0;
+
+	do {
+		// Single event (update cgroup.procs), block indefinitely
+		ready = epoll_wait(epollfd, &event, 1, -1);
```
Comment (Snaipe) [inline outer.c:226]: Check epoll_wait for failures. You'll thank yourself if we ever end up changing the  signal handling, and EINTR starts to sporadically happen.

File: outer.c [hunk 18]
```diff
@@ -183,6 +185,60 @@ static void persist_ns_files(pid_t pid, const char **persist)
...
+
+		// We need a new fd to read from the cgroup.events
+		int readfd = openat(cgroupfd, "cgroup.events", O_RDONLY);
+		read(readfd, &populated, BUFSIZ);
+		close(readfd);
+
+		// Check if there are no procs within the bst cgroup -- if so, delete it
+		if (strncmp(populated, "populated 0", 11) == 0) {
```
Comment (Snaipe) [inline outer.c:234]: The order in which keys appear in the file is not specified. It's more work, but I'd rather you process this file line by line, and search for `populated 0`.

fdopen, fseek, and fgets are your friends. I'd also assume that no line is larger than BUFSIZ characters -- you can error out if fgets returns sizeof (populated) and the last character is not a newline.

File: outer.c [hunk 19]
```diff
@@ -207,6 +263,21 @@ static void persist_ns_files(pid_t pid, const char **persist)
...
+
+	// If cleanup is needed, fork process to epoll cgroup.events
+	if (helper->cleanfd != -1) {
+		pid_t pid = fork();
+		if (pid == -1) {
+			err(1, "outer_helper: cgroup cleanup fork");
+		}
+
+		if (pid == 0) {
+			cgroup_helper(helper->cleanfd, rootpid);
```
Comment (Snaipe) [inline outer.c:276]: cgroup_helper should be ignoring SIGHUP, or calling setsid. This gives the helper a chance to clean up even if the controlling terminal dies.

File: test/bst.t [hunk 20]
```diff
@@ -248,52 +248,42 @@ Testing Environment
...
 
 Test trying to apply climits without specifying a cgroup
 
 	$ ! bst --no-cgroup-remount --limit cpu.max=3333 cat /sys/fs/cgroup/bst/cpu.max
 	bst: unable to apply limits without --cgroup specified: Success
 
-Test cpu.max cgroup usage
+Test overspecified --limits
+	$ ! bst --cgroup /sys/fs/cgroup/bst --limit cpu.max=2222 --limit cpu.max=1111
+	bst: double specified --limit argument: Success
```
Comment (Snaipe) [inline test/bst.t:284]: Haha, weird error! This is caused by the error path calling err rather than errx, since  the former adds strerror(errno) to the message.

File: outer.h [hunk 21]
```diff
@@ -24,6 +24,8 @@ struct outer_helper {
 	pid_t pid;
 	int in;
 	int out;
+
+	int cleanfd;
```
Comment (Snaipe) [inline outer.h:28]: Same comment as earlier, cleanfd tells the reader nothing. This is a cgroup fd that only incidentally ends up being cleaned up.

File: test/bst.t [hunk 22]
```diff
@@ -415,17 +365,12 @@ Test pids.max cgroup usage
 	$ ! bst --cgroup /sys/fs/cgroup/bst --no-cgroup-remount --limit pids.max=-1 cat /sys/fs/cgroup/bst/pids.max 2>/dev/null
 
 Test mutliple climits in conjunction
-
-	$ bst --cgroup=/sys/fs/cgroup/bst --no-cgroup-remount --limit cpu.max="5000 7000" --limit cpu.weight=100 --limit memory.max=1G --limit memory.low=1M --limit memory.max=1G --limit memory.swap.max=1G --limit io.max="8:0 rbps=100 wbps=100 riops=100 wiops=100" --limit io.latency="8:0 target=20" --limit pids.max=50 cat /sys/fs/cgroup/bst/cpu.max; cat /sys/fs/cgroup/bst/cpu.weight; cat /sys/fs/cgroup/bst/memory.min; cat /sys/fs/cgroup/bst/memory.low; cat /sys/fs/cgroup/bst/memory.high; cat /sys/fs/cgroup/bst/memory.max; cat /sys/fs/cgroup/bst/memory.swap.high; cat /sys/fs/cgroup/bst/memory.swap.max; cat /sys/fs/cgroup/bst/io.weight; cat /sys/fs/cgroup/bst/io.max; cat /sys/fs/cgroup/bst/io.latency; cat /sys/fs/cgroup/bst/pids.max
+	$ bst --cgroup=/sys/fs/cgroup/bst --no-cgroup-remount --limit cpu.max="5000 7000" --limit cpu.weight=100 --limit memory.low=1M --limit memory.max=1G --limit pids.max=50 --workdir /sys/fs/cgroup/bst cat cpu.max cpu.weight memory.min memory.low memory.high memory.max io.weight pids.max
```
Comment (Snaipe) [inline test/bst.t:368]: I think I'd also rather have this be on multiple lines for readability.

File: test/bst.t (hunk 12 above)
[the remark this answers is not in the document]
Comment (Snaipe) [reply test/bst.t]: The limits should be applied to bst.pid. cgroup namespaces ensures that the delegated root cgroup is not writeable, even by users that would normally have the permission to do so.

File: test/bst.t (hunk 10 above)
Comment (Snaipe) [reply test/bst.t]: What kind of issue? The syntax I've specified above should be working (i.e, use `$` for the start of the script, then `>` for subsequent lines)

File: test/bst.t (hunk 12 above)
[the remark this answers is not in the document]
Comment (colindrewes) [reply test/bst.t]: Understood.

File: enter.c (hunk 13 above)
Comment (colindrewes) [reply enter.c]: Added setsid() to cgroup_helper.

File: outer.c (hunk 14 above)
Comment (colindrewes) [reply outer.c]: Done

File: outer.c (hunk 15 above)
Comment (colindrewes) [reply outer.c]: Done.

File: outer.c (hunk 16 above)
Comment (colindrewes) [reply outer.c]: Added.

File: outer.c (hunk 18 above)
Comment (colindrewes) [reply outer.c]: Now iterating through all lines and checking for "populated 0".

File: test/bst.t (hunk 20 above)
Comment (colindrewes) [reply test/bst.t]: Switched to errx. It appears --rlimit has this issue as well?

File: outer.h (hunk 21 above)
Comment (colindrewes) [reply outer.h]: Switched to socket IPC.

File: test/bst.t (hunk 22 above)
Comment (colindrewes) [reply test/bst.t]: Reformatted tests.

File: outer.c (hunk 19 above)
Comment (colindrewes) [reply outer.c]: Now calling setsid in the cgroup helper.

File: test/bst.t (hunk 20 above)
Comment (Snaipe) [reply test/bst.t]: Likely. Feel free to fix this in a second PR

File: outer.c [hunk 23]
```diff
@@ -219,24 +229,43 @@ static void cgroup_helper(int cleanfd, pid_t rootpid)
...
+				cgroup_clean(cgroupfd, rootpid);
+				close(subcgroupfd);
+				close(cevent);
+				close(cgroupfd);
+				fclose(eventsfp);
+				return;
+			}
+		}
+
+		close(eventsfd);
```
[the remark this answers is not in the document]
Comment (colindrewes) [reply outer.c]: Fixed

File: outer.c [hunk 24]
```diff
@@ -287,6 +301,16 @@ void outer_helper_spawn(struct outer_helper *helper)
 	if (pipe(pipefds_out) == -1) {
 		err(1, "outer_helper: pipe");
 	}
+	enum {
+		SOCKET_PARENT,
+		SOCKET_CHILD,
+	};
+	int fdpair[2];
+	if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, fdpair) == -1) {
```
[the remark this answers is not in the document]
Comment (colindrewes) [reply outer.c]: Done

File: test/bst.t (hunk 20 above)
Comment (colindrewes) [reply test/bst.t]: Will do
